File: editor/src/core/Outdoor3D.cpp
#include "core/Outdoor3D.h"

#include <cstdio>
#include <unordered_map>

namespace mm2 {

namespace {

constexpr int kHoodBytes = 13;

constexpr int8_t kBundleN[6] = {0, 1, -1, 0, 1, 0};
constexpr int8_t kBundleE[6] = {1, 0, 0, 1, 0, -1};
constexpr int8_t kBundleS[6] = {0, -1, 1, 0, -1, 0};
constexpr int8_t kBundleW[6] = {-1, 0, 0, -1, 0, 1};

constexpr int kStepDx[4] = {0, 1, 0, -1};
constexpr int kStepDy[4] = {1, 0, -1, 0};

const int8_t* directionBundle(int facing) {
    static const int8_t* kAll[4] = {kBundleN, kBundleE, kBundleS, kBundleW};
    return kAll[facing & 3];
}

uint8_t facingMask(int facing) {
    static constexpr uint8_t kMask[4] = {0xC0, 0x30, 0x0C, 0x03};
    return kMask[facing & 3];
}

int neighborScreen(const AttribFile& attrib, int screen, int slot) {
    const int n = attrib.screens[static_cast<size_t>(screen)].neighbor(slot);
    return (n >= 0 && n < kMapScreens) ? n : screen;
}

struct ResolvedCell {
    int screen = 0;
    int lx = 0;
    int ly = 0;
};

ResolvedCell resolveCell(const MapFile& map, const AttribFile& attrib, int screen, int x,
                         int y) {
    ResolvedCell r{};
    r.screen = screen;
    r.lx = x;
    r.ly = y;
    if (r.lx < 0) {
        r.screen = neighborScreen(attrib, r.screen, 3);
        r.lx += kMapGridDim;
    } else if (r.lx >= kMapGridDim) {
        r.screen = neighborScreen(attrib, r.screen, 1);
        r.lx -= kMapGridDim;
    }
    if (r.ly < 0) {
        r.screen = neighborScreen(attrib, r.screen, 2);
        r.ly += kMapGridDim;
    } else if (r.ly >= kMapGridDim) {
        r.screen = neighborScreen(attrib, r.screen, 0);
        r.ly -= kMapGridDim;
    }
    (void)map;
    return r;
}

std::array<uint8_t, 256> gTerrainLookup{};
bool gTerrainLookupLoaded = false;

OutdoorBiome biomeFromSurfaceFlag(uint8_t sf) {
    if (sf == 0xCC) return OutdoorBiome::Ocean;
    if (sf == 0x99) return OutdoorBiome::Tundra;
    if (sf == 0xBB) return OutdoorBiome::Swamp;
    return OutdoorBiome::Desert;
}

void refreshOutdoorHood(const OutdoorMapGrid& grid, int px, int py, int facing,
                        std::array<uint8_t, kHoodBytes>& hood) {
    hood.fill(0);
    const int8_t* b = directionBundle(facing);
    const int dx = b[0], dy = b[1];

    auto rowSampler = [&](int sx, int sy, int outOff) {
        int x = sx, y = sy;
        for (int i = 0; i < kOutdoorHoodRowLen; ++i) {
            hood[static_cast<size_t>(outOff + i)] = grid.at(x, y);
            x += dx;
            y += dy;
        }
    };

    rowSampler(px, py, 0);
    rowSampler(px + b[2], py + b[3], 4);
    rowSampler(px + b[4], py + b[5], 8);
}

void processTerrainRows(const std::array<std::array<uint8_t, kOutdoorHoodRowLen>, 3>& rows,
                        uint8_t currentCell, int facing,
                        const std::array<uint8_t, 256>& lookup,
                        std::array<uint8_t, kOutdoorHoodRowLen>& c6,
                        std::array<uint8_t, kOutdoorHoodRowLen>& c2,
                        std::array<uint8_t, kOutdoorHoodRowLen>& be) {
    for (int i = 0; i < kOutdoorHoodRowLen; ++i) {
        c6[static_cast<size_t>(i)] = lookup[rows[0][static_cast<size_t>(i)] & 0x1F];
        c2[static_cast<size_t>(i)] = lookup[rows[1][static_cast<size_t>(i)] & 0x1F];
        be[static_cast<size_t>(i)] = lookup[rows[2][static_cast<size_t>(i)] & 0x1F];
    }

    const uint8_t near = c6[0];
    if (near >= 1 && near <= 3) {
        const uint8_t fwd = facingMask(facing);
        const uint8_t left =
            (fwd == 0xC0) ? static_cast<uint8_t>(0x03) : static_cast<uint8_t>((fwd << 2) & 0xFF);
        const uint8_t right = (fwd == 0x03) ? static_cast<uint8_t>(0xC0) : (fwd >> 2);
        if ((currentCell & left & 0x55) != 0) c2[0] = near;
        if ((currentCell & right & 0x55) != 0) be[0] = near;
        if ((currentCell & fwd & 0x55) == 0) c6[0] = 0;
    }
}

std::unordered_map<uint8_t, OutdoorBiome> buildSectorLabelBiomes(const AttribFile& attrib) {
    std::unordered_map<uint8_t, OutdoorBiome> out;
    out[0xC3] = OutdoorBiome::Ocean;
    for (const AttribScreen& rec : attrib.screens) {
        if (!rec.isOutside()) continue;
        const uint8_t label = rec.outsideLabel();
        if (out.find(label) != out.end()) continue;
        out[label] = biomeFromSurfaceFlag(rec.outsideFlag());
    }
    return out;
}

OutdoorBiome biomeForCell(uint8_t mapByte, uint8_t sectorLabel,
                          const std::unordered_map<uint8_t, OutdoorBiome>& sectorTable) {
    const uint8_t tid = mapByte & 0x1F;
    if (tid == 2 || tid == 3) return remapOutdoorBiome(OutdoorBiome::Ocean, sectorLabel);
    auto it = sectorTable.find(sectorLabel);
    const OutdoorBiome base =
        (it != sectorTable.end()) ? it->second : OutdoorBiome::Desert;
    return remapOutdoorBiome(base, sectorLabel);
}

void columnBiomes(const OutdoorMapGrid& grid, int px, int py, int facing,
                  const AttribFile& attrib,
                  const std::unordered_map<uint8_t, OutdoorBiome>& sectorTable,
                  std::array<OutdoorBiome, kOutdoorLaneCols>& out) {
    const int8_t* b = directionBundle(facing);
    const int dx = b[0], dy = b[1];
    for (int col = 0; col < kOutdoorLaneCols; ++col) {
        const int wx = px + col * dx;
        const int wy = py + col * dy;
        const int sid = grid.screenIdAt(wx, wy);
        const uint8_t cell = grid.at(wx, wy);
        const uint8_t label = attrib.screens[static_cast<size_t>(sid)].outsideLabel();
        out[static_cast<size_t>(col)] = biomeForCell(cell, label, sectorTable);
    }
}

uint8_t horizonIndex(uint8_t terrainClass) {
    if (terrainClass == 0 || terrainClass > 3) return kOutdoorHorizonNone;
    return static_cast<uint8_t>(terrainClass - 1);
}

int horizonStartCol(const std::array<uint8_t, kOutdoorLaneCols>& l1) {
    for (int col = 0; col < kOutdoorLaneCols; ++col) {
        if (l1[static_cast<size_t>(col)] != kOutdoorHorizonNone) return col;
    }
    return kOutdoorLaneCols;
}

void appendHorizon(std::vector<OutdoorSpriteBlit>& blits, uint8_t idx, int frame, int x, int y) {
    if (idx == kOutdoorHorizonNone || idx > 2) return;
    blits.push_back({OutdoorSpriteBlit::Kind::Horizon,
                     OutdoorBiome::Desert,
                     static_cast<OutdoorHorizonSheet>(idx),
                     frame,
                     x,
                     y});
}

void buildDecorBlits(const std::array<uint8_t, kOutdoorLaneCols>& c6,
                     const std::array<uint8_t, kOutdoorLaneCols>& laneL2,
                     const std::array<uint8_t, kOutdoorLaneCols>& laneL3,
                     const std::array<OutdoorBiome, kOutdoorLaneCols>& colBiomes,
                     std::vector<OutdoorSpriteBlit>& blits) {
    using T = OutdoorHorizonTables;
    for (int col = kOutdoorLaneCols - 1; col >= 0; --col) {
        const size_t ci = static_cast<size_t>(col);
        const OutdoorBiome biome = colBiomes[ci];
        const bool mainA = c6[ci] > 3;
        const bool mainB = laneL2[ci] > 3;
        const bool mainC = laneL3[ci] > 3;
        const int y = T::kDecorY[col];

        auto pushDecor = [&](int frame, int x) {
            blits.push_back(
                {OutdoorSpriteBlit::Kind::Decor, biome, OutdoorHorizonSheet::Outdoor1, frame, x, y});
        };

        if (mainA) {
            pushDecor(col, T::kDecorX);
            if (mainB) pushDecor(col + 12, T::kDecorX);
            if (mainC) pushDecor(col + 16, T::kDecorXBde[col] - kView3DOriginX);
        } else {
            if (mainB) pushDecor(col + 4, T::kDecorX);
            if (mainC) pushDecor(col + 8, T::kDecorX112);
        }
    }
}

void buildHorizonBlits(std::array<uint8_t, kOutdoorLaneCols> c6,
                       std::array<uint8_t, kOutdoorLaneCols> laneL2,
                       std::array<uint8_t, kOutdoorLaneCols> laneL3,
                       std::vector<OutdoorSpriteBlit>& blits) {
    using T = OutdoorHorizonTables;
    std::array<uint8_t, kOutdoorLaneCols> l1{}, l2{}, l3{};
    for (int col = 0; col < kOutdoorLaneCols; ++col) {
        const size_t i = static_cast<size_t>(col);
        l1[i] = horizonIndex(c6[i]);
        l2[i] = horizonIndex(laneL2[i]);
        l3[i] = horizonIndex(laneL3[i]);
    }

    const int start = horizonStartCol(l1);
    int pivot = (start == 3 && l1[3] == kOutdoorHorizonNone) ? 3 : start;

    if (start < kOutdoorLaneCols && l1[static_cast<size_t>(start)] != kOutdoorHorizonNone) {
        appendHorizon(blits, l1[static_cast<size_t>(start)], T::kL1Frame[start],
                      T::kL1X[start], T::kL1Y[start]);
    } else if (start == kOutdoorLaneCols) {
        pivot = 3;
    }

    auto passL2 = [&](int col, int piv) {
        const bool special = (col != 0 && col == piv && l1[static_cast<size_t>(col)] != kOutdoorHorizonNone);
        if (special && l2[static_cast<size_t>(col - 1)] != kOutdoorHorizonNone)
            l2[static_cast<size_t>(col)] = kOutdoorHorizonNone;
        if (l2[static_cast<size_t>(col)] == kOutdoorHorizonNone) return;

        int frame = T::kL2Frame[col];
        int x = T::kL2X[col];
        int y = T::kL2Y[col];
        if (special && l2[static_cast<size_t>(piv)] != kOutdoorHorizonNone) {
            frame = T::kL2Frame[piv - 1];
            x = T::kL2X[piv - 1];
            y = T::kL1Y[piv];
        }
        if (col == 1 && l2[0] == kOutdoorHorizonNone) x = 8;
        if (col == 1 || (col == 2 && col == piv)) x = 8;
        appendHorizon(blits, l2[static_cast<size_t>(col)], frame, x, y);
    };

    auto passL3 = [&](int col, int piv) {
        const bool special = (col != 0 && col == piv && l1[static_cast<size_t>(col)] != kOutdoorHorizonNone);
        if (special && l3[static_cast<size_t>(col - 1)] != kOutdoorHorizonNone)
            l3[static_cast<size_t>(col)] = kOutdoorHorizonNone;
        if (l3[static_cast<size_t>(col)] == kOutdoorHorizonNone) return;

        int frame = T::kL3Frame[col];
        int x = T::kL3X[col];
        int y = T::kL3Y[col];
        if (special && l3[static_cast<size_t>(piv)] != kOutdoorHorizonNone) {
            frame = T::kL3Frame[piv - 1];
            x = T::kL3X[piv - 1];
            y = T::kL1Y[piv];
        }
        if (col == 1 && l3[0] == kOutdoorHorizonNone) x = (col == piv) ? 0xB0 : 0x98;
        appendHorizon(blits, l3[static_cast<size_t>(col)], frame, x, y);
    };

    for (int col = pivot; col >= 0; --col) passL2(col, pivot);
    for (int col = pivot; col >= 0; --col) passL3(col, pivot);
}

}  // namespace

const char* outdoorBiomeFilename(OutdoorBiome biome) {
    switch (biome) {
        case OutdoorBiome::Ocean:  return "ocean.32";
        case OutdoorBiome::Tundra: return "tundra.32";
        case OutdoorBiome::Swamp:  return "swamp.32";
        default:                    return "desert.32";
    }
}

const char* outdoorHorizonFilename(OutdoorHorizonSheet sheet) {
    switch (sheet) {
        case OutdoorHorizonSheet::Outdoor2: return "outdoor2.32";
        case OutdoorHorizonSheet::Outdoor3: return "outdoor3.32";
        default:                          return "outdoor1.32";
    }
}

bool loadOutdoorTerrainLookup(const std::string& dataBinPath, std::array<uint8_t, 256>& out) {
    Bytes blob;
    if (!readFile(dataBinPath, blob)) return false;
    constexpr size_t kA4 = 0x7FFE;
    constexpr size_t kOff = kA4 - 0x720A;
    if (blob.size() < kOff + 256) return false;
    std::copy(blob.begin() + static_cast<std::ptrdiff_t>(kOff),
              blob.begin() + static_cast<std::ptrdiff_t>(kOff + 256), out.begin());
    return true;
}

bool initOutdoorTerrainLookup(const std::string& dataBinPath) {
    if (!loadOutdoorTerrainLookup(dataBinPath, gTerrainLookup)) return false;
    gTerrainLookupLoaded = true;
    return true;
}

const std::array<uint8_t, 256>& outdoorTerrainLookup() {
    if (!gTerrainLookupLoaded)
        initOutdoorTerrainLookup("EXTRACTED/ghidra/mm2_data_00.bin");
    return gTerrainLookup;
}

OutdoorBiome remapOutdoorBiome(OutdoorBiome sheet, uint8_t sectorLabel) {
    if (sectorLabel == 0xC3) return sheet;
    switch (sheet) {
        case OutdoorBiome::Desert: return OutdoorBiome::Ocean;
        case OutdoorBiome::Ocean:  return OutdoorBiome::Tundra;
        case OutdoorBiome::Tundra: return OutdoorBiome::Desert;
        default:                  return sheet;
    }
}

int OutdoorMapGrid::screenIdAt(int x, int y) const {
    return resolveCell(map, attrib, screen, x, y).screen;
}

uint8_t OutdoorMapGrid::at(int x, int y) const {
    const ResolvedCell r = resolveCell(map, attrib, screen, x, y);
    if (r.lx < 0 || r.ly < 0 || r.lx >= kMapGridDim || r.ly >= kMapGridDim) return 0;
    if (r.screen < 0 || r.screen >= kMapScreens) return 0;
    return map.screens[static_cast<size_t>(r.screen)].visualAt(r.lx, r.ly);
}

OutdoorScene buildOutdoorScene(const OutdoorMapGrid& grid, int px, int py, int facing,
                               const AttribFile& attrib) {
    std::array<uint8_t, kHoodBytes> hood{};
    refreshOutdoorHood(grid, px, py, facing, hood);

    OutdoorScene scene{};
    for (int i = 0; i < kOutdoorHoodRowLen; ++i) {
        scene.rowsRaw[0][static_cast<size_t>(i)] = hood[static_cast<size_t>(i)];
        scene.rowsRaw[1][static_cast<size_t>(i)] = hood[static_cast<size_t>(4 + i)];
        scene.rowsRaw[2][static_cast<size_t>(i)] = hood[static_cast<size_t>(8 + i)];
    }

    const auto& lookup = outdoorTerrainLookup();
    processTerrainRows(scene.rowsRaw, grid.at(px, py), facing, lookup, scene.laneC6, scene.laneC2,
                       scene.laneBe);

    std::array<uint8_t, kOutdoorLaneCols> laneL2{}, laneL3{};
    for (int i = 0; i < kOutdoorLaneCols; ++i) {
        laneL2[static_cast<size_t>(i)] = scene.laneC2[static_cast<size_t>(i)];
        laneL3[static_cast<size_t>(i)] = scene.laneBe[static_cast<size_t>(i)];
    }

    const auto sectorTable = buildSectorLabelBiomes(attrib);
    columnBiomes(grid, px, py, facing, attrib, sectorTable, scene.columnBiomes);

    std::array<uint8_t, kOutdoorLaneCols> c6{};
    for (int i = 0; i < kOutdoorLaneCols; ++i)
        c6[static_cast<size_t>(i)] = scene.laneC6[static_cast<size_t>(i)];

    buildDecorBlits(c6, laneL2, laneL3, scene.columnBiomes, scene.decor);
    buildHorizonBlits(c6, laneL2, laneL3, scene.horizon);
    return scene;
}

void stepOutdoorParty(int facing, int& x, int& y, int& screen, const AttribFile& attrib) {
    x += kStepDx[facing & 3];
    y += kStepDy[facing & 3];
    const AttribScreen& rec = attrib.screens[static_cast<size_t>(screen)];
    if (x < 0) {
        const int n = rec.neighbor(3);
        if (n >= 0 && n < kMapScreens) {
            screen = n;
            x = 15;
        }
    } else if (x >= kMapGridDim) {
        const int n = rec.neighbor(1);
        if (n >= 0 && n < kMapScreens) {
            screen = n;
            x = 0;
        }
    }
    if (y < 0) {
        const int n = rec.neighbor(2);
        if (n >= 0 && n < kMapScreens) {
            screen = n;
            y = 15;
        }
    } else if (y >= kMapGridDim) {
        const int n = rec.neighbor(0);
        if (n >= 0 && n < kMapScreens) {
            screen = n;
            y = 0;
        }
    }
    if (x < 0) x = 0;
    if (y < 0) y = 0;
    if (x >= kMapGridDim) x = kMapGridDim - 1;
    if (y >= kMapGridDim) y = kMapGridDim - 1;
}

void dumpOutdoorTrace(const OutdoorScene& scene, int screen, int x, int y, int facing, FILE* out) {
    static const char* kFace[] = {"N", "E", "S", "W"};
    std::fprintf(out, "screen %d  (%d,%d)  facing %s\n", screen, x, y, kFace[facing & 3]);
    std::fprintf(out, "  terrain C6:");
    for (int i = 0; i < kOutdoorLaneCols; ++i) std::fprintf(out, " %u", scene.laneC6[i]);
    std::fprintf(out, "  C2:");
    for (int i = 0; i < kOutdoorLaneCols; ++i) std::fprintf(out, " %u", scene.laneC2[i]);
    std::fprintf(out, "  BE:");
    for (int i = 0; i < kOutdoorLaneCols; ++i) std::fprintf(out, " %u", scene.laneBe[i]);
    std::fprintf(out, "\n");
    std::fprintf(out, "  decor blits: %zu\n", scene.decor.size());
    for (const OutdoorSpriteBlit& b : scene.decor)
        std::fprintf(out, "    %s f%d @ (%d,%d)\n", outdoorBiomeFilename(b.biome), b.frame, b.x,
                     b.y);
    std::fprintf(out, "  horizon blits: %zu\n", scene.horizon.size());
    for (const OutdoorSpriteBlit& b : scene.horizon)
        std::fprintf(out, "    %s f%d @ (%d,%d)\n", outdoorHorizonFilename(b.horizon), b.frame,
                     b.x, b.y);
}

}  // namespace mm2

File: wiki/maze-walker/outdoor3d.js
/** Outdoor first-person scene — ports tools/view3d_outdoor.py. */
"use strict";

import { MAP_GRID, MAP_PAGE_SIZE, MAP_SCREENS, ORIGIN_X, STEP_DX, STEP_DY } from "./view3d.js?v=20260630-slower-anims-2";

const BUNDLE_N = [0, 1, 255, 0, 1, 0];
const BUNDLE_E = [1, 0, 0, 1, 0, 255];
const BUNDLE_S = [0, 255, 1, 0, 255, 0];
const BUNDLE_W = [255, 0, 0, 255, 0, 1];
const BUNDLES = [BUNDLE_N, BUNDLE_E, BUNDLE_S, BUNDLE_W];

const HORIZON_SHEETS = ["outdoor1_32", "outdoor2_32", "outdoor3_32"];

const MAP_TERRAIN_ID_BIOME = { 2: "ocean_32", 3: "ocean_32" };
const SECTOR_LABEL_BIOME = { 0xc3: "ocean_32" };
const BIOME_BY_AREA = { 41: "swamp_32", 42: "swamp_32", 43: "swamp_32", 44: "swamp_32" };
const BIOME_SHEET_REMAP = { desert_32: "ocean_32", ocean_32: "tundra_32", tundra_32: "desert_32" };

const OUTDOOR_DECOR_Y = [0x80 - 20, 0x80 - 35, 0x80 - 50, 0x80 - 60];
const OUTDOOR_DECOR_X = 8;
const OUTDOOR_DECOR_X_112 = 0x70;
const OUTDOOR_DECOR_X_BDE = [184, 160, 136, 112];

const OUTDOOR_L1_Y = [21, 21, 42, 50];
const OUTDOOR_L1_X = [40, 40, 64, 88];
const OUTDOOR_L1_FRAME = [0, 0, 1, 2];
const OUTDOOR_L2_Y = [36, 46, 50, 58];
const OUTDOOR_L2_X = [8, 16, 32, 88];
const OUTDOOR_L2_FRAME = [4, 5, 2, 3];
const OUTDOOR_L3_Y = [36, 46, 50, 58];
const OUTDOOR_L3_X = [176, 152, 136, 120];
const OUTDOOR_L3_FRAME = [6, 7, 2, 3];

function sb(bundle, i) {
  const v = bundle[i] & 0xff;
  return v > 127 ? v - 256 : v;
}

function facingMask(facing) {
  return [0xc0, 0x30, 0x0c, 0x03][facing & 3];
}

function remapBiome(sheet) {
  return BIOME_SHEET_REMAP[sheet] || sheet;
}

/** Stitched outdoor grid (view3d_outdoor.StitchedMap). */
export class StitchedOutdoor {
  constructor(screens, screenId) {
    this.screen = screenId;
    this.screens = screens;
    this.pageByScreen = { [screenId]: screens[screenId].visual };
    const rec = screens[screenId];
    for (let slot = 0; slot < 4; slot++) {
      const n = this._neighborId(rec, slot);
      if (n >= 0 && n < screens.length) this.pageByScreen[n] = screens[n].visual;
    }
  }

  _neighborId(rec, slot) {
    const n = rec.neighbors[slot];
    return n >= 0 && n < this.screens.length ? n : rec.id;
  }

  _neighborOf(screen, slot) {
    const n = this.screens[screen].neighbors[slot];
    return n >= 0 && n < this.screens.length ? n : screen;
  }

  _resolve(x, y) {
    let screen = this.screen;
    let lx = x;
    let ly = y;
    if (lx < 0) {
      screen = this._neighborOf(screen, 3);
      lx += MAP_GRID;
    } else if (lx >= MAP_GRID) {
      screen = this._neighborOf(screen, 1);
      lx -= MAP_GRID;
    }
    if (ly < 0) {
      screen = this._neighborOf(screen, 2);
      ly += MAP_GRID;
    } else if (ly >= MAP_GRID) {
      screen = this._neighborOf(screen, 0);
      ly -= MAP_GRID;
    }
    const page = this.pageByScreen[screen] || new Array(MAP_PAGE_SIZE).fill(0);
    return { page, screen, lx, ly };
  }

  screenIdAt(x, y) {
    return this._resolve(x, y).screen;
  }

  at(x, y) {
    const { page, lx, ly } = this._resolve(x, y);
    if (lx < 0 || ly < 0 || lx >= MAP_GRID || ly >= MAP_GRID) return 0;
    return page[(ly << 4) | lx];
  }
}

function refreshOutdoorHood(grid, px, py, facing) {
  const hood = new Array(13).fill(0);
  const b = BUNDLES[facing & 3];
  const dx = sb(b, 0);
  const dy = sb(b, 1);

  function row(sx, sy, outOff) {
    let x = sx;
    let y = sy;
    for (let i = 0; i < 5; i++) {
      hood[outOff + i] = grid.at(x, y);
      x += dx;
      y += dy;
    }
  }

  row(px, py, 0);
  row(px + sb(b, 2), py + sb(b, 3), 4);
  row(px + sb(b, 4), py + sb(b, 5), 8);
  return hood;
}

function processTerrainRows(rows, currentCell, facing, lookup) {
  const c6 = rows[0].map((b) => lookup[b & 0x1f]);
  const c2 = rows[1].map((b) => lookup[b & 0x1f]);
  const be = rows[2].map((b) => lookup[b & 0x1f]);
  const near = c6[0];
  if (near >= 1 && near <= 3) {
    const fwd = facingMask(facing);
    const left = fwd === 0xc0 ? 0x03 : (fwd << 2) & 0xff;
    const right = fwd === 0x03 ? 0xc0 : fwd >> 2;
    if ((currentCell & left & 0x55) !== 0) c2[0] = near;
    if ((currentCell & right & 0x55) !== 0) be[0] = near;
    if ((currentCell & fwd & 0x55) === 0) c6[0] = 0;
  }
  return [c6, c2, be];
}

export function buildSectorLabelBiomes(screens) {
  const out = { ...SECTOR_LABEL_BIOME };
  for (const rec of screens) {
    if (!rec.outdoor) continue;
    const label = rec.sector;
    if (label in out) continue;
    const sf = rec.surface !== undefined ? rec.surface : 0;
    if (sf === 0xcc) out[label] = "ocean_32";
    else if (sf === 0x99) out[label] = "tundra_32";
    else if (sf === 0xbb) out[label] = "swamp_32";
    else out[label] = "desert_32";
  }
  return out;
}

function biomeForCell(mapByte, sectorLabel, sectorTable) {
  const tid = mapByte & 0x1f;
  if (tid in MAP_TERRAIN_ID_BIOME) return remapBiome(MAP_TERRAIN_ID_BIOME[tid]);
  return remapBiome(sectorTable[sectorLabel] || "desert_32");
}

function columnBiomes(grid, px, py, facing, screens, sectorTable) {
  const b = BUNDLES[facing & 3];
  const dx = sb(b, 0);
  const dy = sb(b, 1);
  const out = [];
  for (let col = 0; col < 4; col++) {
    const wx = px + col * dx;
    const wy = py + col * dy;
    const sid = grid.screenIdAt(wx, wy);
    const cell = grid.at(wx, wy);
    const label = screens[sid].sector;
    out.push(biomeForCell(cell, label, sectorTable));
  }
  return out;
}

function buildDecorBlits(c6, laneL2, laneL3, colBiomes) {
  const blits = [];
  for (let col = 3; col >= 0; col--) {
    const biome = colBiomes[col] || "desert_32";
    const mainA = c6[col] > 3;
    const mainB = laneL2[col] > 3;
    const mainC = laneL3[col] > 3;
    const y = OUTDOOR_DECOR_Y[col];
    if (mainA) {
      blits.push({ sheet: biome, frame: col, x: OUTDOOR_DECOR_X, y });
      if (mainB) blits.push({ sheet: biome, frame: col + 12, x: OUTDOOR_DECOR_X, y });
      if (mainC)
        blits.push({ sheet: biome, frame: col + 16, x: OUTDOOR_DECOR_X_BDE[col] - ORIGIN_X, y });
    } else {
      if (mainB) blits.push({ sheet: biome, frame: col + 4, x: OUTDOOR_DECOR_X, y });
      if (mainC) blits.push({ sheet: biome, frame: col + 8, x: OUTDOOR_DECOR_X_112, y });
    }
  }
  return blits;
}

function horizonIndex(terrainClass) {
  if (terrainClass <= 0 || terrainClass > 3) return 0xff;
  return terrainClass - 1;
}

function horizonSprite(idx, frame, x, y) {
  if (idx === 0xff || idx < 0 || idx >= HORIZON_SHEETS.length) return null;
  return { sheet: HORIZON_SHEETS[idx], frame, x, y };
}

function buildHorizonBlits(c6, laneL2, laneL3) {
  const l1 = c6.slice(0, 4).map((c) => horizonIndex(c));
  const l2 = laneL2.slice(0, 4).map((c) => horizonIndex(c));
  const l3 = laneL3.slice(0, 4).map((c) => horizonIndex(c));
  let start = 4;
  for (let col = 0; col < 4; col++) {
    if (l1[col] !== 0xff) {
      start = col;
      break;
    }
  }
  let pivot = start === 3 && l1[start] === 0xff ? 3 : start;
  const blits = [];

  if (start < 4 && l1[start] !== 0xff) {
    const b = horizonSprite(l1[start], OUTDOOR_L1_FRAME[start], OUTDOOR_L1_X[start], OUTDOOR_L1_Y[start]);
    if (b) blits.push(b);
  } else if (start === 4) pivot = 3;

  function passL2(col, piv) {
    const special = col !== 0 && col === piv && l1[col] !== 0xff;
    if (special && l2[col - 1] !== 0xff) l2[col] = 0xff;
    if (l2[col] === 0xff) return;
    let frame = OUTDOOR_L2_FRAME[col];
    let x = OUTDOOR_L2_X[col];
    let y = OUTDOOR_L2_Y[col];
    if (special && l2[piv] !== 0xff) {
      frame = OUTDOOR_L2_FRAME[piv - 1];
      x = OUTDOOR_L2_X[piv - 1];
      y = OUTDOOR_L1_Y[piv];
    }
    if (col === 1 && l2[0] === 0xff) x = 8;
    if (col === 1 || (col === 2 && col === piv)) x = 8;
    const b = horizonSprite(l2[col], frame, x, y);
    if (b) blits.push(b);
  }

  function passL3(col, piv) {
    const special = col !== 0 && col === piv && l1[col] !== 0xff;
    if (special && l3[col - 1] !== 0xff) l3[col] = 0xff;
    if (l3[col] === 0xff) return;
    let frame = OUTDOOR_L3_FRAME[col];
    let x = OUTDOOR_L3_X[col];
    let y = OUTDOOR_L3_Y[col];
    if (special && l3[piv] !== 0xff) {
      frame = OUTDOOR_L3_FRAME[piv - 1];
      x = OUTDOOR_L3_X[piv - 1];
      y = OUTDOOR_L1_Y[piv];
    }
    if (col === 1 && l3[0] === 0xff) x = col === piv ? 0xb0 : 0x98;
    const b = horizonSprite(l3[col], frame, x, y);
    if (b) blits.push(b);
  }

  for (let col = pivot; col >= 0; col--) passL2(col, pivot);
  for (let col = pivot; col >= 0; col--) passL3(col, pivot);
  return blits;
}

export function buildOutdoorScene(grid, px, py, facing, screens, terrainLookup) {
  const hood = refreshOutdoorHood(grid, px, py, facing);
  const rows = [hood.slice(0, 5), hood.slice(4, 9), hood.slice(8, 13)];
  const [c6, laneD0, laneCc] = processTerrainRows(rows, grid.at(px, py), facing, terrainLookup);
  const laneL2 = laneD0;
  const laneL3 = laneCc;
  const sectorTable = buildSectorLabelBiomes(screens);
  const colBio = columnBiomes(grid, px, py, facing, screens, sectorTable);
  return {
    decor: buildDecorBlits(c6, laneL2, laneL3, colBio),
    horizon: buildHorizonBlits(c6, laneL2, laneL3),
  };
}

export { STEP_DX, STEP_DY };
